File: code/Python/09_facade.py
# -*- coding: utf-8 -*-
#  @file        - 09_facade.py
#  @author      - wdn (dongnian.wang@outlook.com)
#  @brief       - 外观模式
#  @version     - 0.1
#  @date        - 2022-03-10
#  @copyright   - Copyright (c) 2021 


class Subsystem1:
    """ 子系统角色
    """

    def operation1(self):
        print("Method1() of subsystem 01 is called!")



class Subsystem2:
    """ 子系统角色
    """

    def operation2(self):
        print("Method2() of subsystem 02 is called!")


class Subsystem3:
    """ 子系统角色
    """

    def operation3(self):
        print("Method3() of subsystem 03 is called!")


class Facade:
    """ 外观角色
    """

    def __init__(self):
        self._obj1 = Subsystem1()
        self._obj2 = Subsystem2()
        self._obj3 = Subsystem3()

    def operation(self):
        self._obj1.operation1()
        self._obj2.operation2()
        self._obj3.operation3()


def main():
    facade = Facade()
    facade.operation()


if __name__ == "__main__":
    main()

File: code/Cpp/09_facade.cpp
/**
 * @file        - 09_facade.cpp
 * @author      - wdn (dongnian.wang@outlook.com)
 * @brief       - 外观模式
 * @version     - 0.1
 * @date        - 2022-05-07
 * 
 * @copyright Copyright (c) 2022 
 * 
 */

#include <iostream>
#include <string>
#include <memory>

/**
 * 子系统角色
*/
class SubSystem01
{
public:
    void method1()
    {
        std::cout << "Method1() of subsystem 01 is called!" << std::endl;
    }
};

/**
 * 子系统角色
*/
class SubSystem02
{
public:
    void method2()
    {
        std::cout << "Method2() of subsystem 02 is called!" << std::endl;
    }
};

/**
 * 子系统角色
*/
class SubSystem03
{
public:
    void method3()
    {
        std::cout << "Method3() of subsystem 03 is called!" << std::endl;
    }
};

/**
 * 外观角色
*/
class Facade
{
public:
    Facade()
    {
        m_obj1 = std::make_shared<SubSystem01>();
        m_obj2 = std::make_shared<SubSystem02>(); 
        m_obj3 = std::make_shared<SubSystem03>(); 
    }
    void method()
    {
        m_obj1->method1();
        m_obj2->method2();
        m_obj3->method3();
    }
private:
    std::shared_ptr<SubSystem01> m_obj1;
    std::shared_ptr<SubSystem02> m_obj2; 
    std::shared_ptr<SubSystem03> m_obj3; 
};

int main()
{
    Facade* f = new Facade();
    f->method();
    
    delete f;

    return 0;
}
